modules/core/src/main/java/org/apache/ignite/internal/binary/BinaryWriterExImpl.java

```diff
@@ -1880,6 +1892,72 @@ public void writeFieldIdNoSchemaUpdate(int fieldId) {
         fieldCnt++;
     }
 
+    /**
+     * Write byte array from the InputStream with the specified length.
+     *
+     * @param in InputStream.
+     * @param len Length of data in the stream.
+     * @return Number of bytes written.
+     */
+    public int writeByteArrayFromInputStream(InputStream in, int len) throws IOException {
+        out.unsafeEnsure(1 + 4 + len);
+
+        return doWriteByteArrayFromInputStream(in, len);
+    }
+
+    /**
+     * Write byte array from the InputStream with uknown length.
+     *
+     * @param in InputStream.
+     * @return Number of bytes written or -1 if stream contains more than {@code MAX_ARRAY_SIZE} bytes.
+     */
+    public int writeByteArrayFromInputStream(InputStream in) throws IOException {
+        out.unsafeEnsure(1 + 4);
+
+        int writtenLen = doWriteByteArrayFromInputStream(in, MAX_ARRAY_SIZE);
+
+        if (writtenLen == MAX_ARRAY_SIZE && in.read() != -1)
```

Comment:
This method writes `InputStream` from the user and can block forever on `in.read()` invocation.

Reply:
Generally any call to InputStream in all other new stream-related methods can block as well.

I think it's not a problem if method is called from the Thin JDBC Driver only (which is true for now). Since it's a user application thread would block.

On the side if (in future) some server code would call these methods it may be a problem.

***

Reading asyncroniously with timeout looks an overkill for me for the Thin JDBC driver case.  May be just take these methods outside of the BinaryWriterExImpl to JDBC-specific part (to  SqlListenerUtils).

Initially I did so but decided that it's better to have the serialization code at the same file.

***
What is your preference for this?

Reply:
> Generally any call to InputStream in all other new stream-related methods can block as well.

Why do we need to read extra byte here, in the first place?
My point is that `in.read() != -1` are useless here.
I think we must  handle "too much data" case differently - throw from the `doWriteByteArrayFromInputStream` maybe?

Reply:
This is a case when user didn't provide the length of data in the stream vai the API.

So we accept any actual length <= `MAX_ARRAY_SIZE` and thus need to check that there is no more data in stream if `MAX_ARRAY_SIZE` was achieved.

In any case whether here or in `doWriteByteArrayFromInputStream` we still need to do extra `in.read()` to be sure that no more data present in stream.

Moreove in the `doWriteByteArrayFromInputStream` it's supposed that we are not allowed to read more bytes (and move stream pointer) to hadle case what used does pass the length of data.

Or may be I just do not see the obvious solution ?

Reply:
1. To throw exception to user saying that streams longer then `MAX_ARRAY_SIZE` are not supported. Instead of silently strip the tail.
2. Hm, not sure how to answer - I see no other way

Reply:
> To throw exception to user saying that streams longer then MAX_ARRAY_SIZE are not supported. Instead of silently strip the tail.

We can do it when we read more then `MAX_ARRAY_SIZE` bytes from stream while writing data inside `doWriteByteArrayFromInputStream`.

Reply:
Either read 1 byte and expect -1, or read array and expect zero. In any case extra call.

Reply:
I think we can not.  

Since `doWriteByteArrayFromInputStream` is supposed to not read more than limit. No single byte.

I can create another function which would be almost the same but allowed to (try) read more data then limit. In this function the extra read call will only be performed if actual data length is multiple of DEFAULT_BUFFER_SIZE.

 `doWriteByteArrayFromInputStream`  will be used for case for known (declared) length

New  function will be used for case if length is not provided.

Reply:
> Since doWriteByteArrayFromInputStream is supposed to not read more than limit. No single byte.

But you will read extra byte `if (writtenLen == MAX_ARRAY_SIZE && in.read() != -1)` here.

Reply:
only in case if user didn't provide stream length via API call
so we are free to read as we want

Reply:
Looks like with this version we can still have one method and fail if read more then `MAX_ARRAY_SIZE` from InputStream. Logic relies on `-1` as unlimited read flag.

```
    /**
     * Write byte array from the InputStream with uknown length.
     *
     * @param in InputStream.
     * @return Number of bytes written or -1 if stream contains more than {@code MAX_ARRAY_SIZE} bytes.
     */
    public int writeByteArrayFromInputStream(InputStream in) throws IOException {
        out.unsafeEnsure(1 + 4);

        return doWriteByteArrayFromInputStream(in, -1);
    }

    /**
     * Write byte array from the InputStream. No more than {@code limit} bytes will be written.
     *
     * @param in InputStream.
     * @param limit Max length of data to be read from the stream.
     * @return Number of bytes written.
     * @throws IOException If an I/O error occurs.
     */
    private int doWriteByteArrayFromInputStream(InputStream in, int limit) throws IOException {
        out.unsafeWriteByte(GridBinaryMarshaller.BYTE_ARR);

        int lengthPos = out.position();

        out.position(lengthPos + 4);

        int written = 0;

        byte[] buf = new byte[Math.min(limit, DEFAULT_BUFFER_SIZE)];

        while (true) {
            int read = limit > 0
                ? in.read(buf, 0, Math.min(buf.length, limit - written))
                : in.read(buf, 0, buf.length);

            if (read == -1)
                break;

            if (read + written > MAX_ARRAY_SIZE)
                throw new IgniteException("Too much data. Can't write more then " + MAX_ARRAY_SIZE + " bytes from stream");

            out.writeByteArray(buf, 0, read);

            written += read;

            if (written >= limit)
                break;
        }

        out.position(lengthPos);
        out.unsafeWriteInt(written);
        out.position(out.position() + written);

        return written;
    }
```

Reply:
Okay